test/math/ring_signature.cpp

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright (c) 2011-2017 libbitcoin developers (see AUTHORS)
+ *
+ * This file is part of libbitcoin.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <random>
+#include <boost/test/unit_test.hpp>
+#include <bitcoin/bitcoin.hpp>
+
+using namespace bc;
+
+data_chunk new_seed()
+{
+    data_chunk seed(10);
+    pseudo_random_fill(seed);
+    return seed;
+}
+
+ec_secret new_key()
+{
+    const data_chunk seed = new_seed();
+    const wallet::hd_private key(seed);
+    return key.secret();
+}
+
+template <typename SecretsIterator>
+point_list extract_public_keys(SecretsIterator start, SecretsIterator end)
+{
+    point_list ring;
+    for (auto it = start; it != end; ++it)
+    {
+        ec_compressed public_key;
+        bool rc = secret_to_public(public_key, *it);
+        BITCOIN_ASSERT(rc);
+        ring.push_back(public_key);
+    }
+    return ring;
+}
+
+bool is_secret_in_ring(const ec_secret& secret, const point_list& ring)
+{
+    ec_compressed public_key;
+    bool rc = secret_to_public(public_key, secret);
+    BITCOIN_ASSERT(rc);
+    for (const auto& ring_key: ring)
+    {
+        if (ring_key == public_key)
+            return true;
+    }
+    return false;
+}
+
+BOOST_AUTO_TEST_SUITE(ring_signature_tests)
+
+BOOST_AUTO_TEST_CASE(ring_signature__basic_test)
```

Comment:
One giant test case is not ideal. Each individual function/method should be tested independently for each of its conditions. That allows us to see exactly what is broken when a test fails. This is okay as an additional test if it's meant to duplicate a reference implementation.

Test cases should be fully declarative, without internal loops or conditions. There are a few exceptions in libbitcoin, some for historical reasons. 

See also the style/objective of test naming: https://github.com/libbitcoin/libbitcoin/wiki/Test-Guide